pyccel/parser/semantic.py

```diff
@@ -495,6 +496,7 @@ def _infere_type(self, expr, **settings):
             d_var['datatype'   ] = expr.dtype
             d_var['memory_handling'] = 'heap' if expr.rank > 0 else 'stack'
             d_var['memory_location'] = expr.memory_location
+            d_var['current_context'] = expr.current_context
```

Comment:
Do you think this option should be added throughout this function?

Reply:
To be honest i think that the current_context should be checked in semantic, in a similar way to how `PyccelFunctionDef` are visited, i was going to ask for you opinion on that.

Reply:
I'm not sure what you mean

Reply:
I'm not sure if i responded to your interrogation then, where should it be added instead in your opinion ?

Reply:
My question was whether every `if` branch of `_infere_type` should set the `current_context` key.

I would still like to know what you were considering though

Reply:
It's only needed for `CUDA` arrays i don't think its relevant for the others.
I was considering that maybe the way i'm handling `current_context` now is not optimal, maybe if we could "visit" the `CUDA` objects that need the current context (to raise errors or to allocate in the right manner), i would be better instead of adding if statements in function that are not supposed to handle those cases, i'll probably discuss that with @bauom during our next meeting.

Reply:
> It's only needed for CUDA arrays i don't think its relevant for the others.

I'm not sure I agree. It is relevant for `Variable` creation. However I think you are right that the others don't need to specify it and can rely on the default value.

Visiting the CUDA objects definitely sounds like a good idea. I agree that it would be better to have a `_visit_CupyArray` function etc. I'm not sure what the best way to handle the `NumpyArray` case would be though. I suspect the most obvious would be to add a condition in `_assign_lhs_variable`:
```python
if 'device' in self.scope.decorators or 'kernel' in self.scope.decorators and get(d_var, 'memory_location', 'host') == 'host':
                        errors.report("Host arrays cannot be allocated on the Device",
                            symbol = expr,
                            severity = 'fatal')
```

That way you also handle other array creation functions and are reasonably future-proof